src/Analyzers/Core/Analyzers/NamespaceMatchFolder/AbstractNamespaceMatchFolderDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,196 @@
+                s_localizableInsideMessage)
+        {
+        }
+    }
+
+    internal abstract class AbstractNamespaceMatchFolderDiagnosticAnalyzer<TNamespaceSyntax> : AbstractNamespaceMatchFolderDiagnosticAnalyzer
+        where TNamespaceSyntax : SyntaxNode
+    {
+        private static readonly SymbolDisplayFormat s_namespaceDisplayFormat = SymbolDisplayFormat
+            .FullyQualifiedFormat
+            .WithGlobalNamespaceStyle(SymbolDisplayGlobalNamespaceStyle.Omitted);
+
+        /// <summary>
+        /// Gets the language specific syntax facts
+        /// </summary>
+        protected abstract ISyntaxFacts GetSyntaxFacts();
+
+        /// <summary>
+        /// Gets the syntax representing the name of a namespace declaration
+        /// </summary>
+        protected abstract SyntaxNode GetNameSyntax(TNamespaceSyntax namespaceDeclaration);
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected void AnalyzeNamespaceNode(SyntaxNodeAnalysisContext context)
+        {
+            // It's ok to not have a rootnamespace property, but if it's there we want to use it correctly
+            context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(RootNamespaceOption, out var rootNamespace);
+
+            // Project directory is a must to correctly get the relative path and construct a namespace
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(ProjectDirOption, out var projectDir)
+                || string.IsNullOrEmpty(projectDir))
+            {
+                return;
+            }
+
+            if (context.Node is TNamespaceSyntax namespaceDecl)
+            {
+                var currentNamespace = GetNamespaceName(context.SemanticModel, namespaceDecl);
+
+                if (IsFileAndNamespaceMismatch(namespaceDecl, rootNamespace, projectDir, currentNamespace, out var targetNamespace) &&
+                    IsFixSupported(context.SemanticModel, namespaceDecl, context.CancellationToken))
+                {
+                    var nameSyntax = GetNameSyntax(namespaceDecl);
+
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        Descriptor,
+                        nameSyntax.GetLocation(),
+                        additionalLocations: null,
+                        properties: ImmutableDictionary<string, string?>.Empty.Add(TargetNamespace, targetNamespace),
+                        messageArgs: new[] { currentNamespace, targetNamespace }));
+                }
+            }
+
+            static string GetNamespaceName(SemanticModel semanticModel, SyntaxNode namespaceSyntax)
+            {
+                var symbol = semanticModel.GetDeclaredSymbol(namespaceSyntax);
+                RoslynDebug.AssertNotNull(symbol);
+
+                return symbol.ToDisplayString(s_namespaceDisplayFormat);
+            }
```

Comment:
this local function is only called once and doesn't seem to ahve value.  can you just inline this?